core/src/test/java/feign/client/AbstractClientTest.java

```diff
@@ -0,0 +1,254 @@
+package feign.client;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import feign.Client;
+import feign.Feign;
+import feign.FeignException;
+import feign.Headers;
+import feign.Logger;
+import feign.Param;
+import feign.RequestLine;
+import feign.Response;
+import feign.assertj.MockWebServerAssertions;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+
+import static java.util.Arrays.asList;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
+
+import static feign.Util.UTF_8;
+
+/**
+ * {@link AbstractClientTest} can be extended to run a set of tests against any {@link Client} implementation.
+ */
+public abstract class AbstractClientTest {
+
+    @Rule
+    public final ExpectedException thrown = ExpectedException.none();
+    @Rule
+    public final MockWebServer server = new MockWebServer();
+
+    public abstract Client getClient();
```

Comment:
note.. might be easier to detect subtle accidents by exposing `newBuilder()` instead of `getClient()`. Then, you know any call to `Feign.builder()` potentially isn't using the intended client.